stellargraph/data/explorer.py

```diff
@@ -1021,93 +756,137 @@ def _check_parameter_values(self, nodes, n, n_size, graph_schema, seed):
+    It can be used to extract a random sub-graph starting from a set of initial nodes.
+    """
 
-        for d in n_size:
-            if type(d) != int:
-                raise ValueError(
-                    "({}) The neighbourhood size, n_size, must be list of integers.".format(
-                        type(self).__name__
-                    )
-                )
-            if d < 0:
-                raise ValueError(
-                    "({}) n_sie should be positive integer or 0.".format(
-                        type(self).__name__
-                    )
-                )
+    def __init__(self, graph, graph_schema=None, seed=None):
+        super().__init__(graph, graph_schema, seed)
+        if not graph.is_directed():
+            self._raise_error("Graph must be directed")
+
+    def run(self, nodes=None, n=1, in_size=None, out_size=None, seed=None):
+        """
+        Performs a sampled breadth-first walk starting from the root nodes.
+
+        Args:
+            nodes:  <list> A list of root node ids such that from each node n BFWs will be generated up to the
+            given depth d.
+            n: <int> Number of walks per node id.
+            in_size: <list> The number of in-directed nodes to sample with replacement at each depth of the walk.
+            out_size: <list> The number of out-directed nodes to sample with replacement at each depth of the walk.
+            seed: <int> Random number generator seed; default is None
 
-        if graph_schema is not None and type(graph_schema) is not GraphSchema:
-            raise ValueError(
-                "({}) The parameter graph_schema should be either None or of type GraphSchema.".format(
-                    type(self).__name__
-                )
-            )
 
-        if seed is not None:
-            if type(seed) != int:
-                raise ValueError(
-                    "({}) The random number generator seed value, seed, should be integer type or None.".format(
-                        type(self).__name__
+        Returns:
+            A list of multi-hop neighbourhood samples. Each sample expresses multiple undirected walks, but the in-node
+            neighbours and out-node neighbours are sampled separately. Each sample has the format:
+
+                [[node]
+                 [in_1...in_n]  [out_1...out_m]
+                 [in_1.in_1...in_n.in_p] [in_1.out_1...in_n.out_q]
+                    [out_1.in_1...out_m.in_p] [out_1.out_1...out_m.out_q]
+                 [in_1.in_1.in_1...in_n.in_p.in_r] [in_1.in_1.out_1...in_n.in_p.out_s] ...
+                 ...]
+
+            where a single, undirected walk might be, for example:
+
+                [node out_i  out_i.in_j  out_i.in_j.in_k ...]
+        """
+        rs = self._check_parameter_values(nodes, n, in_size, out_size, seed)
+        max_hops = len(in_size)
+        max_slots = 2 ** (max_hops + 1) - 1
+
+        samples = []
+
+        for node in nodes:  # iterate over root nodes
```

Comment:
There's a lot of commonality here with `SampledBreadthFirstWalk`. Do you think it's reasonable to try to factor that out?

Reply:
Yes and no - the output data structure is different and the sequence of sampling is different. But on the other hand, yes, my preference is to refactor code duplication/commonality into modular reuse. Sadly, the code base is a complete mess of copy-pasta and lack of design.